### DISCUSSION TITLE
Drone Client not working for multiple drones in Matlab

### DISCUSSION THREAD

**Discussion by Siddharth2511** [Cosys-Lab/Cosys-AirSim#32] [Q&A] [answered]
I want to move two drones, but it seems that only the second drone (the later for which client is defined) responds to any API calls. 

Here's my Matlab code:

% Define client for Evader and Pursuer drones
evader_name = "Drone1"; % Evader
pursuer_name = "Drone2"; % Pursuer

% Create instances of AirSimClient for both drone
evaderClient = AirSimClient(IsDrone=true, ApiControl=true, IP="127.0.0.1", port=41451, VehicleName=evader_name);
pursuerClient = AirSimClient(IsDrone=true, ApiControl=true, IP="127.0.0.1", port=41451, VehicleName=pursuer_name);

% Take off both drones
evaderClient.drone_client.takeoffAsync().join();
pursuerClient.drone_client.takeoffAsync().join();

% Set initial positions (example coordinates)
evaderClient.drone_client.moveToPositionAsync(10, 10, -10, 5).join(); % Evader moves to (0, 0, -10)
pursuerClient.drone_client.moveToPositionAsync(10, 14, -10, 5).join(); % Pursuer starts at (5, 5, -10)

% Land both drones after chase is complete
evaderClient.drone_client.landAsync().join();
pursuerClient.drone_client.landAsync().join();

This is settings.json:

{
  "SeeDocsAt": "https://github.com/Microsoft/AirSim/blob/main/docs/settings.md",
  "SettingsVersion": 2,
  "SimMode": "Multirotor",
  "Port": 41451,

  "Vehicles": {
    "Drone1": {
      "VehicleType": "SimpleFlight",
      "AllowAPIAlways": true,
      "X": 0,
      "Y": 0,
      "Z": -5,
      "Yaw": 0
    },
    "Drone2": {
      "VehicleType": "SimpleFlight",
      "AllowAPIAlways": true,
      "X": 0,
      "Y": 4,
      "Z": -5,
      "Yaw": 0
    }
  }
}

Only Drone2 responds to any API calls, Drone1 does not even takes off or enters hover mode.

**Comment by WouterJansen** [comment]
I will look into this for a future release. Thank you.

**Comment by WouterJansen** [comment]
I got it to work now (see [commit](https://github.com/Cosys-Lab/Cosys-AirSim/commit/fe0a647f76194f9a9290e7cd103271e4471fd490)) with some slightly updated and fixed matlab code as well. I will look into implementing some form of .join() as well for matlab and make a new release client for matlab once that is implemented and let you know. 

```matlab
evader_name = "Drone1"; % Evader
pursuer_name = "Drone2"; % Pursuer

airSimClient = AirSimClient(IsDrone=true, IP="127.0.0.1", port=41451);

airSimClient.setEnableApiControl("Drone1");
airSimClient.setEnableApiControl("Drone2");

airSimClient.takeoffAsync("Drone1");
airSimClient.takeoffAsync("Drone2");

airSimClient.moveToPositionAsync(10, 10, -10, 5, 3e+38, AirSimDrivetrainTypes.MaxDegreeOfFreedom, true, 0, -1, 1, "Drone1"); % Evader moves to (0, 0, -10)
airSimClient.moveToPositionAsync(10, 14, -10, 5, 3e+38, AirSimDrivetrainTypes.MaxDegreeOfFreedom, true, 0, -1, 1, "Drone2"); % Pursuer starts at (5, 5, -10)

airSimClient.landAsync("Drone1", 360);
airSimClient.landAsync("Drone2", 360);
```

**Reply by Siddharth2511** [reply]
I ran the code after importing changes (changed my unreal environment Plugins). I also uninstalled and reinstalled cosys-airsim matlab addon.
Upon running the aforementioned code, I get the follwoing error message:

Error using AirSimClient/setEnableApiControl
Too many input arguments.

Error in two_drone_api (line 6)
airSimClient.setEnableApiControl("Drone1");

**Reply by WouterJansen** [reply, marked as answer]
That is expected as you would need to use the matlab toolbox from source to see the changes reflected, I had not updated the toolbox itself at that time, just the code.

I have now pushed a new toolbox build 3.1.0 to [file exchange](https://nl.mathworks.com/matlabcentral/fileexchange/168061-cosys-airsim-matlab-api-client).

This includes the previous mentioned fixes as well as the option to do something similar as _.join()_ in Python by setting the _join_ of the async functions to true to wait for function completion. 

This new build also updates the example code and adds a new example for two drones:

```matlab
airSimClient = AirSimClient(IsDrone=true, IP="127.0.0.1", port=41451);

airSimClient.setEnableApiControl("Drone1");
airSimClient.setEnableApiControl("Drone2");

airSimClient.setEnableDroneArm("Drone1");
airSimClient.setEnableDroneArm("Drone2");

airSimClient.takeoffAsync("Drone1", 20, true);
airSimClient.takeoffAsync("Drone2", 20, false);

airSimClient.moveToPositionAsync(10, 10, -5, 5, 3e+38, AirSimDrivetrainTypes.MaxDegreeOfFreedom, true, 0, -1, 1, "Drone1", true);
airSimClient.moveToPositionAsync(10, 14, -5, 5, 3e+38, AirSimDrivetrainTypes.MaxDegreeOfFreedom, true, 0, -1, 1, "Drone2", true); 

airSimClient.landAsync("Drone1", 60, true);
airSimClient.landAsync("Drone2", 60, false);

airSimClient.setDisableDroneArm("Drone1");
airSimClient.setDisableDroneArm("Drone2");

airSimClient.setDisableApiControl("Drone1");
airSimClient.setDisableApiControl("Drone2");
```

This works best with a setting file like this:

```json
{
    "SeeDocsAt": "https://cosys-lab.github.io/settings/",
    "SettingsVersion": 2,
    "ClockSpeed": 1,
    "LocalHostIp": "127.0.0.1",
    "ApiServerPort": 41451,
    "RpcEnabled": true,
    "SimMode": "Multirotor",
    "Vehicles": {
        "Drone1": {
            "VehicleType": "SimpleFlight",
            "AllowAPIAlways": true,
            "X": 0,
            "Y": 0,
            "Z": 0,
            "Yaw": 0
        },
        "Drone2": {
            "VehicleType": "SimpleFlight",
            "AllowAPIAlways": true,
            "X": 5,
            "Y": 0,
            "Z": 0,
            "Yaw": 0
        }
    }
}

```
